src/s2geography/geoarrow.cc

```diff
@@ -742,6 +742,232 @@ void Reader::ReadGeography(const ArrowArray* array, int64_t offset,
+  ImportOptions options_;
+  GeoArrowArrayWriter writer_;
+  GeoArrowVisitor visitor_;
+  GeoArrowCoordView coords_view_;
+  GeoArrowError error_;
+
+  int VisitPoints(const PointGeography& point) {
+    coords_view_.n_coords = 1;
+    coords_view_.n_values = 2;
+    coords_view_.coords_stride = 2;
+    double coords[2];
+    coords_view_.values[0] = &coords[0];
+    coords_view_.values[1] = &coords[1];
+
+    if (point.Points().size() == 0) {
+      // empty Point
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_start(
+          &visitor_, GEOARROW_GEOMETRY_TYPE_POINT, GEOARROW_DIMENSIONS_XY));
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else if (point.Points().size() == 1) {
+      // Point
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_start(
+          &visitor_, GEOARROW_GEOMETRY_TYPE_POINT, GEOARROW_DIMENSIONS_XY));
+      S2LatLng ll(point.Points()[0]);
+      coords[0] = ll.lng().degrees();
+      coords[1] = ll.lat().degrees();
+      GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else {
+      // MultiPoint
+      GEOARROW_RETURN_NOT_OK(
+          visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_MULTIPOINT,
+                              GEOARROW_DIMENSIONS_XY));
+
+      for (const S2Point& pt : point.Points()) {
+        GEOARROW_RETURN_NOT_OK(
+            visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_POINT,
+                                GEOARROW_DIMENSIONS_XY));
+        S2LatLng ll(pt);
+        coords[0] = ll.lng().degrees();
+        coords[1] = ll.lat().degrees();
+        GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+        GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+      }
+
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+    }
+    return GEOARROW_OK;
+  }
+
+  int VisitPolylines(const PolylineGeography& geog) {
+    coords_view_.n_coords = 1;
+    coords_view_.n_values = 2;
+    coords_view_.coords_stride = 2;
+    double coords[2];
+    coords_view_.values[0] = &coords[0];
+    coords_view_.values[1] = &coords[1];
+
+    if (geog.Polylines().size() == 0) {
+      // empty LineString
+      GEOARROW_RETURN_NOT_OK(
+          visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_LINESTRING,
+                              GEOARROW_DIMENSIONS_XY));
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else if (geog.Polylines().size() == 1) {
+      // LineString
+      GEOARROW_RETURN_NOT_OK(
+          visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_LINESTRING,
+                              GEOARROW_DIMENSIONS_XY));
+
+      const auto& poly = geog.Polylines()[0];
+
+      for (int i = 0; i < poly->num_vertices(); i++) {
+        S2LatLng ll(poly->vertex(i));
+        coords[0] = ll.lng().degrees();
+        coords[1] = ll.lat().degrees();
+        GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+      }
```

Comment:
Currently for geometries with >1 coordinates (here for the example of a linestring), I loop over the s2points and call the `coords()` callback multiple times. 

The alternative is to build up a single `GeoArrowCoordView` with the `n_coords` set to something larger than 1, and then only do a single call to `coords()`. 

Now, given that we don't start from a contiguous array of coords, I suppose it will (performance wise) not matter that much if we would do the conversion to a single array ourselves and do a single `coords()` call, or or call `coords()` multiple times and let the underlying writer do the conversion? 
Especially since we don't know for sure if the output actually needs single buffers (like for WKT/WKB it would iterate over the coords pairs anyway then, when passing a single CoordView). 
Unless the typicaly `coords()` visitor implementation might have a bunch of overhead per call? In that case it might be beneficial to limit the number of calls to it (I suppose I should just test the difference .. ;))

Reply:
OK, then I think I will leave it at a `coords()` call per point for this PR, and we can optimize that later